core/src/main/java/org/apache/calcite/sql/type/SqlTypeCoercionRule.java

```diff
@@ -202,7 +202,8 @@ private SqlTypeCoercionRule(Map<SqlTypeName, ImmutableSet<SqlTypeName>> map) {
         coerceRules.copyValues(SqlTypeName.BOOLEAN)
             .add(SqlTypeName.CHAR)
             .add(SqlTypeName.VARCHAR)
-            .addAll(SqlTypeName.NUMERIC_TYPES)
+            .addAll(SqlTypeName.INT_TYPES)
+            .addAll(SqlTypeName.APPROX_TYPES)
```

Comment:
`SqlTypeName.APPROX_TYPES` include `FLOAT`, `REAL` and `DOUBLE`. Are we sure casting these types into boolean always work?
It's strange, it seems that `SELECT CAST(CAST(0 AS float) AS BOOLEAN)` works fine (returns `false`). But if we try a more "complex" query: `select cast(cast(p0 as float) as boolean) from (values (0)) as t(p0)` ; it will fail at runtime with `CompileException: Line 21, Column 31: Cannot cast "double" to "boolean"`.
This type of query will be generated if we add these tests into `SqlOperatorBaseTest#testCastToBoolean`:
```
tester.checkBoolean("cast(cast(0 as float) as boolean)", Boolean.FALSE);
tester.checkBoolean("cast(cast(0 as real) as boolean)", Boolean.FALSE);
tester.checkBoolean("cast(cast(0 as double) as boolean)", Boolean.FALSE);
```
which, if I am not mistaken, will fail if we add them to `SqlOperatorBaseTest#testCastToBoolean`. Therefore, perhaps we should consider `FLOAT`, `REAL` and `DOUBLE` (together with `DECIMAL`) as non-castable into boolean?

Reply:
About this bug I create a JIRA yesterday(in https://issues.apache.org/jira/browse/CALCITE-4810) before this pr.
I tested the sql "tester.checkBoolean("cast(cast(0.0 as double) as boolean)", Boolean.FALSE);" in master where i didn't apply my change and it threw the same exception just like you says.
I tried to find the reason and add the message into [CALCITE-4810]. It seems the bug that these two behavoirs codegen and constant simplifying is not unified.
I think your suggestion that avoids casting from float, real, double and decimal to boolean is reasonable and this also can solve this problem of inconsistent behavior. And more importantly, it follows the SQL standard.

Reply:
IMO it makes sense, if we say that `DECIMAL` cannot be cast into `BOOLEAN`; consider the same for `FLOAT`, `REAL` and `DOUBLE`, I think it is a consistent approach.
However, if we confirm to go in this direction:
- It seems some test will need adjustments
- The cast matrix in `reference.md` will need to be modified for those rows too.